### PR TITLE
more precise use of timers

### PR DESCRIPTION
**Pull request by ZappoMan**
- cleans up our use of QTimer to use precision timers for things that need to be precise
- improves Script timers to fire even if their interval is less than the script update rate
- properly punishes timers and update methods if they are taking too long to execute

### CODE REVIEW

**Comment by ZappoMan** [comment]
Test Plan:
- smoke test, make sure everything seems to work ok.
- test a script set interval to see that under 250ms, it's more accurate. In particular timers < 30ms will be much more accurate.

**Comment by hifi-gustavo** [comment]
Build console output for commit: a72f60152befa0c6ecf4841b106d21b420ea0d5b available at: https://gist.github.com/56594d498ba957b7059aed11068aed4b<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9283-a72f601.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9283-a72f601.dmg

**Comment by OmegaHeron** [comment]
Testing Windows 10
Smoke tests pass.

I made a quick and dirty script to run 100 ticks at various intervals and give average event fire time then ran it 5 times.  Yes, this PR is more accurate with short fire times down to minimum event time of ~16.67ms.  As an ancient it's always "fun" to see that 16.67ms wall that was built with the original IBM PC system IRQ clock all the way back in the '80s when a 4.77MHz CPU with 256K RAM was uber.

---

This PR

With 100 ms I see;
`[11/29 13:27:58] [tickTest.js] For 100 samples average interval = 100 ms`
`[11/29 13:28:12] [tickTest.js] For 100 samples average interval = 100.01 ms`
`[11/29 13:28:23] [tickTest.js] For 100 samples average interval = 100.01 ms`
`[11/29 13:28:35] [tickTest.js] For 100 samples average interval = 100 ms`
`[11/29 13:28:46] [tickTest.js] For 100 samples average interval = 100.01 ms`

With 50 ms I see;
`[11/29 13:31:32] [tickTest.js] For 100 samples average interval = 50.18 ms`
`[11/29 13:31:38] [tickTest.js] For 100 samples average interval = 50.01 ms`
`[11/29 13:31:44] [tickTest.js] For 100 samples average interval = 50.01 ms`
`[11/29 13:31:50] [tickTest.js] For 100 samples average interval = 50 ms`
`[11/29 13:31:57] [tickTest.js] For 100 samples average interval = 50 ms`

With 25 ms I see;
`[11/29 13:33:2] [tickTest.js] For 100 samples average interval = 25.17 ms`
`[11/29 13:33:6] [tickTest.js] For 100 samples average interval = 25.01 ms`
`[11/29 13:33:9] [tickTest.js] For 100 samples average interval = 25.17 ms`
`[11/29 13:33:13] [tickTest.js] For 100 samples average interval = 25.01 ms`
`[11/29 13:33:16] [tickTest.js] For 100 samples average interval = 25 ms`

With 10ms I see;
`[11/29 13:35:20] [tickTest.js] For 100 samples average interval = 16.34 ms`
`[11/29 13:35:23] [tickTest.js] For 100 samples average interval = 16.68 ms`
`[11/29 13:35:25] [tickTest.js] For 100 samples average interval = 16.34 ms`
`[11/29 13:35:28] [tickTest.js] For 100 samples average interval = 16.01 ms`
`[11/29 13:35:30] [tickTest.js] For 100 samples average interval = 16.67 ms`

With 5ms I see;
`[11/29 13:36:42] [tickTest.js] For 100 samples average interval = 16.34 ms`
`[11/29 13:36:46] [tickTest.js] For 100 samples average interval = 16.34 ms`
`[11/29 13:36:48] [tickTest.js] For 100 samples average interval = 16.66 ms`
`[11/29 13:36:51] [tickTest.js] For 100 samples average interval = 16.67 ms`
`[11/29 13:36:55] [tickTest.js] For 100 samples average interval = 16.34 ms`

With 18ms I see;
[11/29 13:38:28] [tickTest.js] Set interval = 18
`[11/29 13:38:28] [tickTest.js] For 100 samples average interval = 18.01 ms`
`[11/29 13:38:31] [tickTest.js] For 100 samples average interval = 18 ms`
`[11/29 13:38:33] [tickTest.js] For 100 samples average interval = 18 ms`
`[11/29 13:38:36] [tickTest.js] For 100 samples average interval = 18.17 ms`
`[11/29 13:38:38] [tickTest.js] For 100 samples average interval = 18.18 ms`

With 17ms I see;
`[11/29 13:39:34] [tickTest.js] For 100 samples average interval = 17 ms`
`[11/29 13:39:36] [tickTest.js] For 100 samples average interval = 17.17 ms`
`[11/29 13:39:39] [tickTest.js] For 100 samples average interval = 17.01 ms`
`[11/29 13:39:41] [tickTest.js] For 100 samples average interval = 17.17 ms`
`[11/29 13:39:43] [tickTest.js] For 100 samples average interval = 17.01 ms`

With 16ms I see;
`[11/29 13:40:58] [tickTest.js] For 100 samples average interval = 16.67 ms`
`[11/29 13:41:0] [tickTest.js] For 100 samples average interval = 16.67 ms`
`[11/29 13:41:3] [tickTest.js] For 100 samples average interval = 16.5 ms`
`[11/29 13:41:5] [tickTest.js] For 100 samples average interval = 16.51 ms`
`[11/29 13:41:8] [tickTest.js] For 100 samples average interval = 16.68 ms`

---

Comparing to Master (5850)

`[11/29 13:43:44] [tickTest.js] Set interval = 100`
`[11/29 13:43:44] [tickTest.js] For 100 samples average interval = 109.34 ms`
`[11/29 13:43:56] [tickTest.js] For 100 samples average interval = 109.35 ms`
`[11/29 13:44:10] [tickTest.js] For 100 samples average interval = 109.35 ms`
`[11/29 13:44:23] [tickTest.js] For 100 samples average interval = 109.34 ms`
`[11/29 13:44:34] [tickTest.js] For 100 samples average interval = 109.34 ms`

`[11/29 13:45:47] [tickTest.js] Set interval = 50`
`[11/29 13:45:47] [tickTest.js] For 100 samples average interval = 62.51 ms`
`[11/29 13:45:54] [tickTest.js] For 100 samples average interval = 62.51 ms`
`[11/29 13:46:1] [tickTest.js] For 100 samples average interval = 62.51 ms`
`[11/29 13:46:8] [tickTest.js] For 100 samples average interval = 62.51 ms`
`[11/29 13:46:17] [tickTest.js] For 100 samples average interval = 62.5 ms`

`[11/29 13:47:3] [tickTest.js] Set interval = 25`
`[11/29 13:47:3] [tickTest.js] For 100 samples average interval = 31.34 ms`
`[11/29 13:47:7] [tickTest.js] For 100 samples average interval = 31.34 ms`
`[11/29 13:47:10] [tickTest.js] For 100 samples average interval = 31.34 ms`
`[11/29 13:47:14] [tickTest.js] For 100 samples average interval = 31.18 ms`
`[11/29 13:47:18] [tickTest.js] For 100 samples average interval = 31.35 ms`

`[11/29 13:48:27] [tickTest.js] Set interval = 10`
`[11/29 13:48:27] [tickTest.js] For 100 samples average interval = 16.67 ms`
`[11/29 13:48:30] [tickTest.js] For 100 samples average interval = 16.68 ms`
`[11/29 13:48:32] [tickTest.js] For 100 samples average interval = 16.67 ms`
`[11/29 13:48:34] [tickTest.js] For 100 samples average interval = 16.67 ms`
`[11/29 13:48:37] [tickTest.js] For 100 samples average interval = 16.67 ms`

File: interface/src/ui/DiskCacheEditor.cpp [hunk 1]
```diff
@@ -91,7 +91,7 @@ void DiskCacheEditor::makeDialog() {
 
     static const int REFRESH_INTERVAL = 100; // msec
     _refreshTimer = new QTimer(_dialog);
-    _refreshTimer->setInterval(REFRESH_INTERVAL);
+    _refreshTimer->setInterval(REFRESH_INTERVAL); // 100ms, Qt::CoarseTimer acceptable
```
Comment (jherico) [inline interface/src/ui/DiskCacheEditor.cpp]: You say here 100ms... coarse is acceptable, but elsewhere in the code you set the cutoff as 200 ms.

**Comment by OmegaHeron** [comment]
A few more with this PR, couple longer durations and testing the area around 200ms.  So tick < 200ms = precise it seems which lines up with CR comment.

`Set interval = 1000`
`For 100 samples average interval = 1000.03 ms`

`Set interval = 2750`
`For 100 samples average interval = 2750.08 ms`

`Set interval = 221`
`For 100 samples average interval = 234.34 ms`

`Set interval = 199`
`For 100 samples average interval = 199.01 ms`

`Set interval = 201`
`For 100 samples average interval = 203.18 ms`

`Set interval = 200`
`For 100 samples average interval = 203.18 ms`

File: interface/src/ui/DiskCacheEditor.cpp (hunk 1 above)
Comment (ZappoMan) [inline interface/src/ui/DiskCacheEditor.cpp]: I will update the comment, this particular timer probably doesn't really need 100ms accuracy... it appears to be an obscure UI that shows the current state of your disk cache, not something that updates at a real time frequency.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 4b9345d7c6eda6358dc98923a3420280cdf7f88e available at: https://gist.github.com/56594d498ba957b7059aed11068aed4b<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9283-4b9345d.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9283-4b9345d.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 6c86675f24fa5c46badbafe6e9aa854b1bc9e4c6 available at: https://gist.github.com/56594d498ba957b7059aed11068aed4b<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9283-6c86675.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9283-6c86675.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 73cb100d7690bc915760f2e73a72ea1984077cb3 available at: https://gist.github.com/56594d498ba957b7059aed11068aed4b<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9283-73cb100.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9283-73cb100.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 1bf39652235f84b28fcb008671fcb54e9e59934b available at: https://gist.github.com/56594d498ba957b7059aed11068aed4b<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9283-1bf3965.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9283-1bf3965.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 65c8d4b16b5513f513c9a61da6b1fa22829e9e48 available at: https://gist.github.com/56594d498ba957b7059aed11068aed4b<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9283-65c8d4b.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9283-65c8d4b.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: b1d556aee2efd01a2b4815d49dfd079d68e0e6d8 available at: https://gist.github.com/56594d498ba957b7059aed11068aed4b<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9283-b1d556a.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9283-b1d556a.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 206660cecdc8b19bd6f94ec3fe43e00b25213ab1 available at: https://gist.github.com/56594d498ba957b7059aed11068aed4b<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9283-206660c.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9283-206660c.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 1fcfc9444b200d0707fceed4152c31af1473077e available at: https://gist.github.com/56594d498ba957b7059aed11068aed4b<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9283-1fcfc94.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9283-1fcfc94.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 6892e8761b545f1def4f272b621905311e8a20b7 available at: https://gist.github.com/56594d498ba957b7059aed11068aed4b<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9283-6892e87.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9283-6892e87.dmg

**Comment by OmegaHeron** [comment]
Windows 10 - i5 6500 CPU - GTX970 GPU

Ran through test script list and looks like my results match expectation... results below;

Interval 20, TIMER_WORK_EFFORT 999, UPDATE_WORK_EFFORT 999

`[11/30 12:25:48] [tickTest2.js] Set interval = 20`
`[11/30 12:25:48] [tickTest2.js] For 10 samples average update = 15 ms average work:0 ms`
`[11/30 12:25:49] [tickTest2.js] For 10 samples average update = 16.7 ms average work:0.1 ms`
`[11/30 12:25:49] [tickTest2.js] For 10 samples average update = 16.6 ms average work:0 ms`
`[11/30 12:25:49] [tickTest2.js] For 10 samples average update = 16.7 ms average work:0 ms`
`[11/30 12:25:49] [tickTest2.js] For 10 samples average update = 16.7 ms average work:0 ms`
`[11/30 12:25:49] [tickTest2.js] For 10 samples average update = 16.7 ms average work:0.2 ms`
`[11/30 12:25:49] [tickTest2.js] For 10 samples average update = 16.6 ms average work:0 ms`
`[11/30 12:25:50] [tickTest2.js] For 10 samples average update = 16.7 ms average work:0.1 ms`
`[11/30 12:25:50] [tickTest2.js] For 10 samples average update = 16.6 ms average work:0 ms`
`[11/30 12:25:50] [tickTest2.js] For 10 samples average update = 16.7 ms average work:0 ms`
`[11/30 12:25:50] [tickTest2.js] For 10 samples average update = 16.7 ms average work:0 ms`
`[11/30 12:25:50] [tickTest2.js] For 10 samples average update = 16.6 ms average work:0 ms`
`[11/30 12:25:50] [tickTest2.js] For 100 samples average interval = 20.01 ms average work:0.04 ms`

Interval 5, TIMER_WORK_EFFORT 999, UPDATE_WORK_EFFORT 999

`[11/30 12:28:6] [tickTest2.js] Set interval = 5`
`[11/30 12:28:6] [tickTest2.js] For 10 samples average update = 15.1 ms average work:0.1 ms`
`[11/30 12:28:6] [tickTest2.js] For 10 samples average update = 16.6 ms average work:0 ms`
`[11/30 12:28:6] [tickTest2.js] For 10 samples average update = 16.8 ms average work:0.1 ms`
`[11/30 12:28:6] [tickTest2.js] For 100 samples average interval = 5.01 ms average work:0.04 ms`

Interval 11, TIMER_WORK_EFFORT 999, UPDATE_WORK_EFFORT 999

`[11/30 12:29:54] [tickTest2.js] Set interval = 11`
`[11/30 12:29:54] [tickTest2.js] For 10 samples average update = 15.1 ms average work:0.1 ms`
`[11/30 12:29:54] [tickTest2.js] For 10 samples average update = 16.6 ms average work:0 ms`
`[11/30 12:29:54] [tickTest2.js] For 10 samples average update = 16.7 ms average work:0 ms`
`[11/30 12:29:55] [tickTest2.js] For 10 samples average update = 16.7 ms average work:0.1 ms`
`[11/30 12:29:55] [tickTest2.js] For 10 samples average update = 16.6 ms average work:0 ms`
`[11/30 12:29:55] [tickTest2.js] For 10 samples average update = 16.7 ms average work:0 ms`
`[11/30 12:29:55] [tickTest2.js] For 100 samples average interval = 11.01 ms average work:0.06 ms`

Interval 11, TIMER_WORK_EFFORT 100051, UPDATE_WORK_EFFORT 999

`[11/30 12:32:23] [tickTest2.js] Set interval = 11`
`[11/30 12:32:23] [tickTest2.js] For 10 samples average update = 15 ms average work:0.1 ms`
`[11/30 12:32:23] [tickTest2.js] For 10 samples average update = 16.7 ms average work:0 ms`
`[11/30 12:32:23] [tickTest2.js] For 10 samples average update = 16.7 ms average work:0 ms`
`[11/30 12:32:23] [tickTest2.js] For 10 samples average update = 16.6 ms average work:0 ms`
`[11/30 12:32:23] [tickTest2.js] For 10 samples average update = 16.7 ms average work:0 ms`
`[11/30 12:32:24] [tickTest2.js] For 10 samples average update = 16.6 ms average work:0 ms`
`[11/30 12:32:24] [tickTest2.js] For 100 samples average interval = 11.04 ms average work:3.75 ms`

Interval 11, TIMER_WORK_EFFORT 500022, UPDATE_WORK_EFFORT 999

`[11/30 12:34:6] [tickTest2.js] Set interval = 11`
`[11/30 12:34:6] [tickTest2.js] For 10 samples average update = 18.2 ms average work:0 ms`
`[11/30 12:34:6] [tickTest2.js] For 10 samples average update = 19.2 ms average work:0 ms`
`[11/30 12:34:6] [tickTest2.js] For 10 samples average update = 19 ms average work:0 ms`
`[11/30 12:34:6] [tickTest2.js] For 10 samples average update = 18.9 ms average work:0 ms`
`[11/30 12:34:7] [tickTest2.js] For 10 samples average update = 19.3 ms average work:0 ms`
`[11/30 12:34:7] [tickTest2.js] For 10 samples average update = 19.1 ms average work:0 ms`
`[11/30 12:34:7] [tickTest2.js] For 10 samples average update = 19.1 ms average work:0 ms`
`[11/30 12:34:7] [tickTest2.js] For 10 samples average update = 19.2 ms average work:0.1 ms`
`[11/30 12:34:7] [tickTest2.js] For 10 samples average update = 19.1 ms average work:0.1 ms`
`[11/30 12:34:7] [tickTest2.js] For 10 samples average update = 19.5 ms average work:0.1 ms`
`[11/30 12:34:7] [tickTest2.js] For 100 samples average interval = 19.25 ms average work:18.91 ms`

Interval 40, TIMER_WORK_EFFORT 1000099, UPDATE_WORK_EFFORT 999

`[11/30 12:36:18] [tickTest2.js] Set interval = 40`
`[11/30 12:36:18] [tickTest2.js] For 10 samples average update = 27.9 ms average work:0 ms`
`[11/30 12:36:19] [tickTest2.js] For 10 samples average update = 36.5 ms average work:0 ms`
`[11/30 12:36:19] [tickTest2.js] For 10 samples average update = 37.4 ms average work:0 ms`
`[11/30 12:36:19] [tickTest2.js] For 10 samples average update = 37.3 ms average work:0 ms`
`[11/30 12:36:20] [tickTest2.js] For 10 samples average update = 38 ms average work:0 ms`
`[11/30 12:36:20] [tickTest2.js] For 10 samples average update = 37.3 ms average work:0.1 ms`
`[11/30 12:36:21] [tickTest2.js] For 10 samples average update = 37.8 ms average work:0 ms`
`[11/30 12:36:21] [tickTest2.js] For 10 samples average update = 37.7 ms average work:0.1 ms`
`[11/30 12:36:21] [tickTest2.js] For 10 samples average update = 38.2 ms average work:0.1 ms`
`[11/30 12:36:22] [tickTest2.js] For 10 samples average update = 37.6 ms average work:0.1 ms`
`[11/30 12:36:22] [tickTest2.js] For 10 samples average update = 37.5 ms average work:0 ms`
`[11/30 12:36:22] [tickTest2.js] For 100 samples average interval = 40.69 ms average work:37.94 ms`

Interval 10, TIMER_WORK_EFFORT 1007, UPDATE_WORK_EFFORT 1000010

`[11/30 12:39:30] [tickTest2.js] Set interval = 10`
`[11/30 12:39:31] [tickTest2.js] For 10 samples average update = 64.8 ms average work:37.9 ms`
`[11/30 12:39:31] [tickTest2.js] For 10 samples average update = 73.5 ms average work:38 ms`
`[11/30 12:39:32] [tickTest2.js] For 100 samples average interval = 18.4 ms average work:0.06 ms`

Interval 10, TIMER_WORK_EFFORT 1000021, UPDATE_WORK_EFFORT 1000012

`[11/30 12:42:5] [tickTest2.js] Set interval = 10`
`[11/30 12:42:6] [tickTest2.js] For 10 samples average update = 102.6 ms average work:41.1 ms`
`[11/30 12:42:7] [tickTest2.js] For 10 samples average update = 110.6 ms average work:36.7 ms`
`[11/30 12:42:8] [tickTest2.js] For 10 samples average update = 110.9 ms average work:36.6 ms`
`[11/30 12:42:9] [tickTest2.js] For 10 samples average update = 111.5 ms average work:36.6 ms`
`[11/30 12:42:10] [tickTest2.js] For 10 samples average update = 111.6 ms average work:36.7 ms`
`[11/30 12:42:12] [tickTest2.js] For 10 samples average update = 111.8 ms average work:37 ms`
`[11/30 12:42:13] [tickTest2.js] For 10 samples average update = 111.5 ms average work:36.7 ms`
`[11/30 12:42:14] [tickTest2.js] For 10 samples average update = 112.2 ms average work:37.1 ms`
`[11/30 12:42:15] [tickTest2.js] For 10 samples average update = 112 ms average work:36.9 ms`
`[11/30 12:42:16] [tickTest2.js] For 100 samples average interval = 109.95 ms average work:38.6 ms`

**Pull request merged**